ci/cd/utils/artifact_repository.py

```diff
@@ -0,0 +1,623 @@
+    :param args: The arguments passed in by the user or derived by the script
+    :param subdir: An optional subdirectory in which to store the files. Post-pended to the end of the prefix.
+    :return: A string containing the S3 key prefix to be used to uploading and downloading files to the artifact repository
+    """
+    prefix = "{git_sha}/{lib_type}/{os}/{variant}/".format(**vars(args))
+    if subdir:
+        return "{}{}/".format(prefix, subdir)
+    return prefix
+
+
+def push_artifact(args: argparse.Namespace):
+    """
+    Pushes the artifact to the artifact repository
+    :param args: The arguments passed in to this script by the user
+    :return 0 for success, non-zero for failure
+    """
+
+    args = probe(args)
+
+    logger.info('Pushing artifact with: ')
+    logger.info('COMMIT ID   : {}'.format(args.git_sha))
+    logger.info('OS          : {}'.format(args.os))
+    logger.info('VARIANT     : {}'.format(args.variant))
+    logger.info("LIBMXNET    : {}".format(args.libmxnet))
+    logger.info("LICENSES    : {}".format(args.licenses))
+    logger.info("DEPENDENCIES: {}".format(args.dependencies))
+    logger.info("")
+
+    # Upload mxnet
+    try:
+        logger.info('Uploading libmxnet library...')
+        s3_upload(args.bucket, get_s3_key_prefix(args), [args.libmxnet])
+        logger.info("")
+
+        # Upload licenses, if necessary
+        if args.licenses:
+            logger.info('Uploading licenses...')
+            s3_upload(args.bucket, get_s3_key_prefix(args, subdir='licenses'), args.licenses)
+            logger.info("")
+
+        # Upload dependencies, if necessary
+        if args.dependencies:
+            logger.info('Uploading dependencies...')
+            s3_upload(args.bucket, get_s3_key_prefix(args, subdir='dependencies'), args.dependencies)
+            logger.info("")
+    except botocore.exceptions.BotoCoreError as e:
+        logger.error('Error uploading artifact')
+        logger.error(e)
+        return 1
+
+    logger.info('Successfully pushed artifact')
+    return 0
+
+
+def pull_artifact(args: argparse.Namespace):
+    """
+    Pulls the artifact from the artifact repository
+    :param args: The arguments passed in to this script by the user
+    :return 0 for success, 1 for unexpected failure, 2 for no artifact found failure
+    """
+    if not args.variant:
+        logger.warning('''variant not set. Using 'cpu' by default.''')
+        args.variant = 'cpu'
+
+    args = probe(args)
+
+    logger.info('Pulling artifact with: ')
+    logger.info('COMMIT ID   : {}'.format(args.git_sha))
+    logger.info('OS          : {}'.format(args.os))
+    logger.info('VARIANT     : {}'.format(args.variant))
+    logger.info('To directory: {}'.format(args.destination))
+
+    try:
+        if not try_s3_download(args.bucket, get_s3_key_prefix(args), args.destination):
+            logging.error('No artifacts found for this configuration.')
+            return 2
```

Comment:
Raise Exception insteads of magic numbers?

Reply:
Should be fixed.